proxy/src/schema.rs

```diff
@@ -44,24 +77,132 @@ impl Into<ProjectId> for IdInput {
     }
 }
 
-/// Encapsulates read paths in API.
-pub struct Query;
+/// Object representing a git branch.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, GraphQLObject)]
+struct Branch {
+    /// Name of the branch.
+    name: String,
+}
 
-#[juniper::object(Context = Context)]
-impl Query {
-    fn apiVersion() -> &str {
-        "1.0"
-    }
+/// Object representing a git tag.
+///
+/// We still need full tag support.
+#[derive(Debug, Eq, Ord, PartialEq, PartialOrd, GraphQLObject)]
+struct Tag {
+    /// Name of the tag.
+    name: String,
+}
 
-    fn projects(ctx: &Context) -> FieldResult<Vec<Project>> {
-        Ok(ctx.source.get_all_projects())
-    }
+/// Representation of a person (e.g. committer, author, signer) from a repository. Usually
+/// extracted from a signature.
+#[derive(GraphQLObject)]
+struct Person {
+    /// Name part of the commit signature commit.
+    name: String,
+    /// Email part of the commit signature commit.
+    email: String,
+    /// Reference (url/uri) to a persons avatar image.
+    avatar: String,
+}
 
-    fn project(ctx: &Context, id: IdInput) -> FieldResult<Option<Project>> {
-        Ok(ctx.source.get_project(id.into()))
+/// Representation of a code commit.
+#[derive(GraphQLObject)]
+struct Commit {
```

Comment:
So actually, this is making think that this module should really be `schema::git` or `git::schema`. Because it's just git related. But there might be things we can extract out such as the user profile, and these things would go under `schema::mod`. What do you think? Obviously that can be done in another PR :)

Reply:
This module is the surface of the GraphQL API the proxy provides. While it could be broken apart potentially, I want to defer that to the latest possible point and until then keep it centralised so a reader can get an effortless overview of what to expect of the endpoint. We haven't found a way yet ot surface that more ellegantly (i.e. a generated schema file from an introspection call).

Reply:
There is definitely room to isolate parts of it, do you see an obvious drawback of having it organised as is?

Reply:
It's more just a "housekeeping" thing. If you separate boundaries early you can notice where things need to overlap and where things can be extended easily. But ya this sort of thing can be done as other factors come into play :)

Reply:
Good point, I gonna attempt this and see what the rippling effects are as the main endpoint `Query` type is an integration point for all these other types. Also curios if it makes the massive test suite that currently starts to grow at the bottom of `schema.rs` becomes more managable.

Reply:
Modularity is King :crown: